compiler/src/dotty/tools/dotc/quoted/PickledQuotes.scala

```diff
@@ -273,10 +273,26 @@ object PickledQuotes {
           unpickler.enter(Set.empty)
 
           val tree = unpickler.tree
-          QuotesCache(pickled) = tree
 
+          var includesSymbolDefinition = false
           // Make sure trees and positions are fully loaded
-          tree.foreachSubTree(identity)
+          new TreeTraverser {
+            def traverse(tree: Tree)(using Context): Unit =
+              tree match
+                case _: DefTree =>
+                  if !tree.symbol.hasAnnotation(defn.QuotedRuntime_SplicedTypeAnnot)
+                  && !tree.symbol.hasAnnotation(defn.QuotedRuntimePatterns_patternTypeAnnot)
+                  then
+                    includesSymbolDefinition = true
+                case _ =>
+              traverseChildren(tree)
+          }.traverse(tree)
```

Comment:
Tas part was inspired by https://github.com/scala/scala3/pull/12242/commits/94f069ea9a7d2de922614f31b0ed823d41929c8d, but instead of looking for any symbol in the unpicked tree to decide whether to cache/uncache, we decide whether to cache with or without the context owner. This keeps the previous performance improvements while keeping the owners unpicked symbols always correct